linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -0,0 +1,2206 @@
+      Nil
+    )
+    addHelperImport(
+      genFunctionID.installJSProperty,
+      List(anyref, anyref, anyref, RefType.funcref, RefType.funcref),
+      Nil
+    )
+    addHelperImport(
+      genFunctionID.installJSStaticProperty,
+      List(anyref, anyref, anyref, RefType.funcref, RefType.funcref),
+      Nil
+    )
+    addHelperImport(
+      genFunctionID.jsSuperSelect,
+      List(anyref, anyref, anyref),
+      List(anyref)
+    )
+    addHelperImport(
+      genFunctionID.jsSuperSelectSet,
+      List(anyref, anyref, anyref, anyref),
+      Nil
+    )
+    addHelperImport(
+      genFunctionID.jsSuperCall,
+      List(anyref, anyref, anyref, anyref),
+      List(anyref)
+    )
+  }
+
+  /** Generates all the non-type definitions of the core Wasm lib. */
+  private def genHelperDefinitions()(implicit ctx: WasmContext): Unit = {
+    genStringLiteral()
+    genCreateStringFromData()
+    genTypeDataName()
+    genCreateClassOf()
+    genGetClassOf()
+    genArrayTypeData()
+    genIsInstance()
+    genIsAssignableFromExternal()
+    genIsAssignableFrom()
+    genCheckCast()
+    genGetComponentType()
+    genNewArrayOfThisClass()
+    genAnyGetClass()
+    genNewArrayObject()
+    genIdentityHashCode()
+    genSearchReflectiveProxy()
+    genArrayCloneFunctions()
+  }
+
+  private def newFunctionBuilder(functionID: FunctionID, originalName: OriginalName)(
+      implicit ctx: WasmContext): FunctionBuilder = {
+    new FunctionBuilder(ctx.moduleBuilder, functionID, originalName, noPos)
+  }
+
+  private def newFunctionBuilder(functionID: FunctionID)(
+      implicit ctx: WasmContext): FunctionBuilder = {
+    newFunctionBuilder(functionID, OriginalName(functionID.toString()))
+  }
+
+  private def genStringLiteral()(implicit ctx: WasmContext): Unit = {
+    val fb = newFunctionBuilder(genFunctionID.stringLiteral)
+    val offsetParam = fb.addParam("offset", Int32)
+    val sizeParam = fb.addParam("size", Int32)
+    val stringIndexParam = fb.addParam("stringIndex", Int32)
+    fb.setResultType(RefType.any)
+
+    val str = fb.addLocal("str", RefType.any)
+
+    fb.block(RefType.any) { cacheHit =>
+      fb += GlobalGet(genGlobalID.stringLiteralCache)
+      fb += LocalGet(stringIndexParam)
+      fb += ArrayGet(genTypeID.anyArray)
+
+      fb += BrOnNonNull(cacheHit)
+
+      // cache miss, create a new string and cache it
+      fb += GlobalGet(genGlobalID.stringLiteralCache)
+      fb += LocalGet(stringIndexParam)
+
+      fb += LocalGet(offsetParam)
+      fb += LocalGet(sizeParam)
+      fb += ArrayNewData(genTypeID.i16Array, genDataID.string)
+      fb += Call(genFunctionID.createStringFromData)
+      fb += LocalTee(str)
+      fb += ArraySet(genTypeID.anyArray)
+
+      fb += LocalGet(str)
+    }
+
+    fb.buildAndAddToModule()
+  }
+
+  /** `createStringFromData: (ref array u16) -> (ref any)` (representing a `string`). */
+  private def genCreateStringFromData()(implicit ctx: WasmContext): Unit = {
```

Comment:
> But why not just emit all the string constants in JS in an array and get them from there?

Because `array` in Wasm is opaque to JS 😞 only `i32`, `i64`, `f32`, `f64` can be exported to JS, and hence iterating over Wasm array and convert to JS char one by one, and concat them is the only way :P

A good news is that, there's a proposal called [js-string-builtins](https://github.com/WebAssembly/js-string-builtins) and it contains `"wasm:js-string" "fromCharCodeArray"` that convert `i16 array` to JS string.
https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#wasmjs-string-fromcharcodearray
This should makes it much easier and faster to create JS string from Wasm array.